### ISSUE TITLE
Empty string in Enum got converted to '0'

### ISSUE THREAD

**Issue by fhewitt** [lukeautry/tsoa#460]
When using empty string `''` in string Enum used a input parameter, it got converted as `'0'`.

So imagine an enum like:

```ts
export enum MyThings {
    Undefined = '',
    Something = 'Something',
    Otherwise = 'Another Something',
```

## Sorting

* **I'm submitting a ...**
  - [X] bug report

* I confirm that I
  - [X] used the [search](https://github.com/lukeautry/tsoa/search?type=Issues) to make sure that a similar issue hasn't already been submit

## Expected Behavior

I would expect that the empty string is returned in the swagger doc.

```yaml
    "MyThings ": {
        "enum": [
            "",
            "Something",
            "Another Something"
        ],
        "type": "string"
    },
```

And the '' value should be accepted during data validation.

## Current Behavior
But I got that in swagger config and for validation:

```yaml
    "MyThings ": {
        "enum": [
            "0",
            "Something",
            "Another Something"
        ],
        "type": "string"
    },
```

Well it's 0 because it's in first position, but moving it elsewhere, it will take the index of the position:

```yaml
    "MyThings ": {
        "enum": [
            "Something",
            "Another Something",
            "2"
        ],
        "type": "string"
    },
```

Trying to pass empty string as the value to end point, I got the error "Should be one of [0, Something, Another Something]"

## Workaround
You can change the empty string for something else, like a single space, dash or the word 'Undefined'. But this break Boolean conversion and seems counter-intuitive.

## Context (Environment)
Version of the library: 2.3.8 & 2.4.11
Version of NodeJS: 10.14.1
* Confirm you were using yarn not npm: [X]

## Breaking change?
This change the current behaviour, so it's a breaking change in principle. But this seems unlikely that anyone have used empty string in enum as it is.

**Comment by fhewitt** [comment]
I'm not familiar enough with this code to propose a PR, but I suspect that the issue is around the typeResolver.ts line 286, where we do `return getEnumValue(member) || String(index);` Because '' is false, this defaults to the index. I would suggest to change for:

```ts
        const enumValue = getEnumValue(member);
        return !!enumValue || enumValue === '' ? enumValue : String(index);
```

It still pass the testing, seem to work with what I read of string enum in TypeScript doc. Unsure if this require additionnal testing (which would probably goes in unit/swagger/definitionsGeneration/definitions.spec.ts, but after that I'm a bit lost on the proper approach).

**Comment by dgreene1** [comment]
Before someone can take this on, can you confirm that an empty string is even a valid enum value in OpenAPI?

Because this seems strange:
```yaml
MyEnum:
  type: string
    - 
    - Something
    - Another Something
```

Since this is a rare use case, I suspect that it won’t get picked up unless you or someone who needs the use case is willing to put in a PR. I apologize for being that direct, but I want to clarify that this is lower priority than some of the larger initiatives that are coming up. And since it seems important to you, I wanted to encourage you to take up the PR.

Note: the workaround is to use an abstraction enum that does not have an empty string value and but is instead a nullable enum.

**Comment by fhewitt** [comment]
My issue is that I have not full control over the data, I don't receive null, but empty string. But I suppose I can temporarily marshalling the data when I get them before I sent them to the API. But I can temper the value before I sent them back to our API.

Seem valid in OpenAPI, even if it could looks a bit weird in documentation.

I push the test further, but even by changing the model to add the empty string in allowed values, it still raise the error at runtime, I suspect that the validation run a "non falsy" check on the value. If I got some more time soon, I will continue the investigation and propose a PR.

**Comment by rgripper** [comment]
Validation also doesn't work properly with empty string enum value.
Could we have this check in `validateEnum` method:
[](https://github.com/lukeautry/tsoa/blob/28e599d44571803e8c90697a0d78b95550d28077/src/routeGeneration/templateHelpers.ts#L240)
changed to:
`if (enumValue === undefined) {`
to

**Comment by dgreene1** [comment]
Sure, please make a PR.

**Comment by dgreene1** [comment]
Released in 2.5.3
